main.py

```diff
@@ -780,19 +827,21 @@ def _install_flows(
                     )
                 flow_mod = build_flow_mod_from_command(flow, command)
                 flow_mod.pack()
-                flow_mods.append(flow_mod)
-                flows.append(flow)
-                owners.append(flow_dict.get("owner", "no_owner"))
-                flow_dicts.append(
+                flow_mods[switch.id].append(flow_mod)
+                flows[switch.id].append(flow)
+                owners[switch.id].append(flow_dict.get("owner", "no_owner"))
+                flows_dict_by_switch[switch.id].append(
                     {"flow": flow_dict, "flow_id": flow.id, "switch": switch.id}
                 )
         if save and command == "add":
-            self.flow_controller.upsert_flows(
-                [flow.match_id for flow in flows], flow_dicts
-            )
+            for switch in switches:
```

Comment:
Here you're performing `O(switches)` bulk upsertions, you can do in a single upsertion, each match.id is already a unique identifier that is a function of the switch.id too

Reply:
Now sending 1 write to DB.

Reply:
Great. Can you refactor this into a single dict indexed by the match.id and then the value is the flow dict? that way it makes clear that no match id is expected to be duplicated making the intension clear, then you pass to upsert_flows the dict.keys() and its dict.values(), see what I mean?

Reply:
Collecting data in a dictionary and send the keys and values now